### PR TITLE
Refactor from_levels_and_colors.

### PR DESCRIPTION
**Pull request by anntzer**
... to use check_in_list, and to derive the number of expected colors
from the slice extents.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/colors.py [hunk 1]
```diff
@@ -2007,33 +2007,23 @@ def from_levels_and_colors(levels, colors, extend='neither'):
...
     n_data_colors = len(levels) - 1
-    n_expected_colors = n_data_colors + extra_colors
-    if len(colors) != n_expected_colors:
-        raise ValueError('With extend == {0!r} and n_levels == {1!r} expected'
-                         ' n_colors == {2!r}. Got {3!r}.'
-                         ''.format(extend, len(levels), n_expected_colors,
-                                   len(colors)))
-
-    cmap = ListedColormap(colors[colors_i0:colors_i1], N=n_data_colors)
+    n_expected = n_data_colors + color_slice.start - (color_slice.stop or 0)
```
Comment (story645) [inline lib/matplotlib/colors.py]: finding this a bit hard to follow ```(color_slice.stop or 0)```
-it always yields 0 or 1 and that gets subtracted?

Comment (anntzer) [reply lib/matplotlib/colors.py]: 0 if stop is None (i.e. it goes all the way to the end), -(-1) if we drop the last element (so we need one more).
